src/RegistryCompilerPass.php

```diff
@@ -58,23 +73,35 @@ private function buildRegistry(ContainerBuilder $container, $registryId, array $
 					}
 				}
 
-				$collection->add($id, $options);
+				$collection->add([$id, $optionsResolver->getMetadata($options)], $options);
 			}
 
 		}
 
 		$targetId = $registryId;
 		$methodName = $tagOptions['method'];
+		$useMetadata = null !== $tagOptions['metadata'];
+
 		if ($tagOptions['use_collection']) {
 			$targetId = "{$registryId}.collection";
-			$container->setDefinition($targetId, new Definition(\ArrayObject::class))->setPublic(false);
+
+			$collectionClass = $useMetadata ? MetaDataIterator::class : \ArrayObject::class;
+			$container->setDefinition($targetId, new Definition($collectionClass))->setPublic(false);
```

Comment:
`use_collection` is a great optimization in terms of the number of calls generated so making a man-in-the-middle class here doesn't fit imho. We've been there in docplanner when we were optimizing service definitions and in many places we dropped such proxy classes (same thing goes with `lazy: true` in service def) because of the memory consumption impact